pennylane/templates/subroutines/permute.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the ``permute`` template.
+"""
+
+import pennylane as qml
+
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+from pennylane.templates.decorator import template
+from pennylane.wires import Wires
+
+
+@template
+def Permute(permutation, wires):
+    r"""Applies a permutation to a set of wires.
+
+    Example usage:
+
+    .. code-block:: python
+
+        import pennylane as qml
+
+        with qml.tape.QuantumTape() as tape:
+            # Send contents of wire 4 to wire 0, of wire 2 to wire 1, etc.
+            qml.templates.Permute([4, 2, 0, 1, 3], wires=[0, 1, 2, 3, 4])
+
+    See "Usage Details" below for further examples.
+
+    Args:
+        permutation (list): A list of wire labels that represents the new ordering of wires
+            after the permutation. The list may consist of integers or strings, so long as
+            they match the labels of ``wires``.
+        wires (Iterable or Wires): Wires that the permutation acts on. Accepts an iterable
+            of numbers or strings, or a Wires object.
+
+    Raises:
+        ValueError: if inputs do not have the correct format
+
+    .. UsageDetails::
+
+        As a simple example, suppose we have a 5-qubit tape with wires labeled
+        by the integers ``[0, 1, 2, 3, 4]``. We apply a permutation to shuffle the
+        order to ``[4, 2, 0, 1, 3]`` (i.e., the qubit state that was previously on
+        wire 4 is now on wire 0, the one from 2 is on wire 1, etc.).
+
+
+        .. code-block:: python
+
+            import pennylane as qml
+            from pennylane.templates import Permute
+
+            with qml.tape.QuantumTape() as tape:
+                # RZs added to ensure numerical ordering in drawing
+                for wire in range(5):
+                    qml.RZ(0, wires=wire)
```

Comment:
Worth making this nonzero? E.g., even just `np.arange()`?

Reply:
We could also remove the "#RZs added..." line, as not sure if it adds much for a user.

Reply:
```suggestion
                    qml.RZ(wire, wires=wire)
```
Could this work? Since it's just an integer

Reply:
Yup - adding the RZs at all is purely just for making sure things get printed out with all the wires, and with them in the expected order.

Reply:
Am going to leave as-is since it's just an example.